creusot-contracts/src/ghost_ptr.rs

```diff
@@ -0,0 +1,165 @@
+    #[requires(self@.is_empty())]
+    pub fn drop(self) {}
+}
+
+impl<T: ?Sized> GhostPtrExt<T> for GhostPtr<T> {
+    #[trusted]
+    #[logic]
+    #[ensures(forall<t: GhostPtrToken<T>> !t@.contains(result))]
+    #[ensures(forall<ptr: GhostPtr<T>> ptr.addr_logic() == result.addr_logic() ==> ptr == result)]
+    fn null_logic() -> Self {
+        absurd
+    }
+
+    #[trusted]
+    #[logic]
+    fn addr_logic(self) -> Int {
+        absurd
+    }
+}
+
+pub trait GhostPtrExt<T: ?Sized>: Sized {
+    #[logic]
+    fn null_logic() -> Self;
+    #[logic]
+    fn addr_logic(self) -> Int;
+}
+
+extern_spec! {
+    impl<T> *const T {
+        // Safety since addr_logic is uninterpreted this just claims ptr::addr is deterministic
+        #[trusted]
+        #[ensures(result@ == self.addr_logic())]
+        fn addr(self) -> usize;
+
+        /// Check if `self` was created with [`GhostPtr::null`]
+        #[trusted]
+        #[ensures(result == (self == GhostPtr::<T>::null_logic()))]
+        fn is_null(self) -> bool;
+    }
+
+    mod std {
+        mod ptr {
+            /// Creates a null [`GhostPtr`] that no GhostToken has permission to
+            // Safety even though this pointer is dangling no GhostToken has permission to it so it's okay
+            #[trusted]
+            #[ensures(result == GhostPtr::<T>::null_logic())]
+            fn null<T>() -> *const T
+            where
+                T: Sized;
+        }
+    }
+}
+
+// Coverts `val` into a [`GhostPtr`] and gives `t` permission to it
+// Safety this pointer was owned by a box so no other GhostToken could have permission to it
+#[trusted]
+#[ensures(!(*t)@.contains(result))]
+#[ensures((^t)@ == (*t)@.insert(result, *val))]
+pub fn ptr_from_box<T>(val: Box<T>, t: &mut GhostPtrToken<T>) -> *const T {
+    Box::into_raw(val)
+}
+
+/// Immutably borrows `self`
+// Safety no other token has permission to `self`
+// `t` cannot be used to mutably borrow `self` as long as the shared lifetime lasts
+#[trusted]
+#[requires(t@.contains(ptr))]
+#[ensures(*result == t@.lookup(ptr))]
+pub fn ptr_as_ref<T>(ptr: *const T, t: &GhostPtrToken<T>) -> &T {
+    unsafe { &*ptr }
+}
+
+#[trusted]
+#[requires(new_model.subset((*t)@))]
+#[ensures((*result)@ == *new_model)]
+pub fn shrink_token_ref<T>(
+    t: &GhostPtrToken<T>,
+    new_model: Ghost<<GhostPtrToken<T> as ShallowModel>::ShallowModelTy>,
+) -> &GhostPtrToken<T> {
+    t
+}
+
+/// Mutably borrows `ptr` and shrinks `t` so that it can no longer be used to access `ptr`
+// Safety no other token has permission to `self`
+// `t` can no longer be used to access `ptr`
+#[trusted]
+#[requires((**t)@.contains(ptr))]
+#[ensures(*result == (**t)@.lookup(ptr))]
+#[ensures((*^t)@ == (**t)@.remove(ptr))]
+#[ensures((^*t)@.remove(ptr) == (^^t)@)]
+#[ensures((^*t)@.get(ptr) == Some((^result).make_sized()))]
+#[ensures(!(^^t)@.contains(ptr))]
+pub fn ptr_as_mut2<'o, 'i, T>(ptr: *const T, t: &'o mut &'i mut GhostPtrToken<T>) -> &'i mut T {
```

Comment:
We have to come up with a better name. Something like `remove_mut_ptr`?

Reply:
I changed the name to `take_mut` (after https://doc.rust-lang.org/std/primitive.slice.html#method.take_mut), are you ok with that?